jupyterlab_server/config.py

```diff
@@ -0,0 +1,229 @@
+# coding: utf-8
+"""JupyterLab Server config"""
+
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
+
+from glob import glob
+import json
+from os.path import join as pjoin
+import os.path as osp
+import os
+
+from jupyter_core.paths import jupyter_path
+from jupyter_server.services.config.manager import ConfigManager, recursive_update
+from traitlets import Bool, HasTraits, List, Unicode, default
+
+from .server import url_path_join as ujoin
+
+
+# -----------------------------------------------------------------------------
+# Module globals
+# -----------------------------------------------------------------------------
+
+DEFAULT_TEMPLATE_PATH = osp.join(osp.dirname(__file__), 'templates')
+
+
+def get_dynamic_extensions(labextensions_path):
+    """Get the metadata about dynamic extensions
+    """
+    dynamic_exts = dict()
+    for ext_dir in labextensions_path:
+        ext_pattern = ext_dir + '/**/package.json'
+        for ext_path in [path for path in glob(ext_pattern, recursive=True)]:
+            with open(ext_path) as fid:
+                data = json.load(fid)
+            if data['name'] not in dynamic_exts:
+                data['ext_dir'] = ext_dir
+                data['ext_path'] = osp.dirname(ext_path)
+                data['is_local'] = False
+                dynamic_exts[data['name']] = data
+    return dynamic_exts
+
+
+def get_static_page_config(app_settings_dir=None, logger=None):
+    # Start with the deprecated `share/jupyter/lab/settings/page_config.json` data
+
+    cm = ConfigManager(config_dir_name="labconfig")
+    page_config = cm.get('page_config')
+
+    # TODO: remove in 4.0
```

Comment:
Let's instead allow passing in a `page_config` file here so it is more generic and won't need to change later.  Let's not warn here either.